casper/src/main/scala/coop/rchain/casper/ReportingCasper.scala

```diff
@@ -186,10 +108,10 @@ object ReportingCasper {
     new ReportingCasper[F] {
       val codecK                                                     = serializeTaggedContinuation.toCodec
       implicit val m: RSpaceMatch[F, BindPattern, ListParWithRandom] = matchListPar[F]
-      val prettyPrinter                                              = RhoPrinter()
-      val transformer                                                = setupTransformer(prettyPrinter)
 
-      override def trace(hash: BlockHash): F[Report] =
+      override def trace(
+          hash: BlockHash
+      ): F[Option[Either[ReplayFailure, List[(ProcessedDeploy, Seq[Seq[ReportingEvent]])]]]] =
```

Comment:
Nested Option and Either are always suspicious. Here both define possible error, so `None` case can also be another `Left` case with some wrapper like `ReportError` which can have cases for block errors and case to wrap `ReplayFailure`. `Report` trait in `ReportingRoutes` can then be simplified because it's doing similar thing.

I'm more for ApplicativeErrors because they are much easier to extend and handle. This wrapper is a small step into this direction.

Reply:
@tgrospic I updated the embedded Option and Either. Take a look and see if that is what you want.